docs/v4/tutorial/first-app.md

```diff
@@ -0,0 +1,1051 @@
+     *
+     * @return int The new ID
+     */
+    public function insertUser(UserCreateData $user): int
+    {
+        $values = [
+            'username' => $user->username,
+            'first_name' => $user->firstName,
+            'last_name' => $user->lastName,
+            'email' => $user->email,
+        ];
+
+        // Insert record
+        $table = $this->queryFactory->table('users');
+        $table->insert($values);
+
+        // Return new primary key value (id)
+        return (int)$table->getLastInsertValue();
+    }
+}
+
+```
+
+Note that we have declared `QueryFactory` as a dependency, because the repository requires a database connection.
+
+You can use the query builder to create `SELECT`, `UPDATE`, `INSERT` and `DELETE` statements.
+Here are some examples:
+
+Select all rows (for simple queries):
+
+```php
+$rows = $this->queryFactory->table('users')->select( /* where */ );
+
+foreach ($rows as $row) {
+    print_r($row);
+}
+```
+
+Select all rows (for complex queries):
+
+```php
+$table = $this->queryFactory->table('users');
+
+$select = $table->getSql()->select();
+$select->columns(['id']);
+$select->where(['id' => 1]);
+
+$rows = $table->selectWith($select);
+
+foreach ($rows as $row) {
+    print_r($row);
+}
+```
+
+Select only the first row:
+
+```php
+$row = $this->queryFactory->table('users')
+    ->select(['id' => 1])->current();
+```
+
+Insert a record:
+
+```php
+$table = $this->queryFactory->table('users');
+$table->insert($values);
+
+$newId = (int)$table->getLastInsertValue();
+```
+
+Update a record:
+
+```php
+$this->queryFactory->table('users')
+    ->update(['email' => 'new@example.com'], ['id' => 1]);
+```
+
+Delete a record:
+
+```php
+$this->queryFactory->table('users')
+    ->delete(['id' => 1]);
+```
+
+The last part is to register a new route for `POST /users`.
+
+Create a new action class in: `src/Action/UserCreateAction.php`:
+
+```php
+<?php
+
+namespace App\Action;
+
+use App\Domain\User\Data\UserCreateData;
+use App\Domain\User\Service\UserCreator;
+use Slim\Http\Response;
+use Slim\Http\ServerRequest;
+
+final class UserCreateAction
+{
+    private $userCreator;
+
+    public function __construct(UserCreator $userCreator)
+    {
+        $this->userCreator = $userCreator;
+    }
+
+    public function __invoke(ServerRequest $request, Response $response): Response
+    {
+        // Collect input from the HTTP request
+        $data = (array)$request->getParsedBody();
+
+        // Mapping (should be done in a mapper class)
+        $user = new UserCreateData();
+        $user->username = $data['username'];
+        $user->firstName = $data['first_name'];
+        $user->lastName = $data['last_name'];
+        $user->email = $data['email'];
+
+        // Invoke the Domain with inputs and retain the result
+        $userId = $this->userCreator->createUser($user);
+
+        // Build the HTTP response
+        return $response->withJson(['user_id' => $userId]);
```

Comment:
I think that a write operation should be separated from a read operation. For such a case I would add a new route like `GET /users/{id}` to read the user by the given id.

Reply:
When an entity is created, the entire entity created should be returned as the response. Returning just the id makes very little sense as from a client perspective you have to complete another operation to read the newly created entity, not to mention you're now having to return a custom single property for every entity type created.

E.g.:
- When creating a user: `{"user_id": 1}`
- When creating a post: `{"post_id": 1}`

That's not very scalable and introduces room for error on both client/server side where as returning the entire entity the identifying property will always have the same format:
- When creating a user: `{"id": 1, "username": "john"}`
- When creating a post: `{"id": 1, "title": "example"}`

It makes it easier from the client side to build a scalable pattern. Imagine having 100 different entities, having to keep track of 100 different property names. That doesn't make much sense.

Also the response code should be `201` when an entity is created successfully

Reply:
I think there should be a “barrier” between source data and output, so schema changes do not affect clients. I would not mix an "entity" with the presentation and transformation for the (JSON) output. These are two different responsibilities for me and they should not be mixed.

Properties names are a matter of taste (e.g. id or user_id). I don't have a problem with properties like `user_id` or `post_id`, because I can see what it is, without guessing. Some people like it more explicit and other not.

The status 201 is correct, I will fix this.

Reply:
> I think there should be a “barrier” between source data and output

Sure, this is exactly why if you're going to return a JSON formatted response you should implement the `JsonSerializable` interface on your entity or a custom serializer that's context bound which enables you to present it differently than the source data.

> so schema changes do not affect clients. I would not mix an "entity" with the presentation and transformation for the (JSON) output.

Schema changes should only happen on API version changes anyway, not sure why that's a concern here.

> Properties names are a matter of taste (e.g. id or user_id). I don't have a problem with properties like user_id or post_id, because I can see what it is, without guessing. Some people like it more explicit and other not.

Again, it's just not a scalable way to do things. The overhead created by these varying property names is of poor design in my opinion. We can agree to disagree.

I also noticed, There's no `User` entity in your skeleton, why? I think DTOs should only be used as an entry point into the application layer. Clearly our opinions also differ on this as you use DTOs through all layers of your application. This is the flow that I think makes the most sense though:

1. Client Requests Entity Creation Endpoint (POST /users)
2. Incoming data goes into a DTO
3. DTO is validated, if invalid the data does not travel any further
4. DTO has been validated, an entity is then created in memory to travel within other layers of your application.
5. The entity can be directly handled by a Domain Service or Repository at this point. Services and Repositories only return valid entities and valid collections of entities.
6. Once the entity is created by the repository, it will append the newly generated identifying property (if using numerical identifiers this is necessary, otherwise UUIDs would be generated when the DTO is formed into a valid entity upon entering the Application layer in step 4)
7. The fully formed valid entity is returned to the application later where it can be serialized and presented to the requester.

Reply:
By definition an entity must have an ID and is more related to the persistence layer (e.g. mapped to table). A DTO is decoupled, doesn't need an id, and can be used to transfer data inside **and** outside the domain. 

JsonSerializable is to bounded to the entity (table structure) and not flexible enough (for me). I don't want to write simple CRUD APIs. I need the full control about the data structure, even very complex and nested structures. I also don't want to manage different API versions (for different structures) within entities. Entities are just not the right "tool" for this task (IMHO) and looks like a poor design.

> 2. Incoming data goes into a DTO
> 3. DTO is validated, if invalid the data does not travel any further

I would validate the raw data (array) first and then create the DTO from it. The reason is, that the incoming data are casted into specific data types. If you validate casted data, your validation does't give you correct results.

> 4. DTO has been validated, an entity is then created in memory to travel within other layers of your application.

DTOs are commonly known as objects passing through the architectural boundaries to transfer data (and not entities).

> 5. The entity can be directly handled by a Domain Service

If you have an entity then you imply that you always have an ID. This makes no sense for me in every case. A DTO is for transfering data inside and outside the domain (not an entity). 

> 6. ... otherwise UUIDs would be generated when the DTO is formed into a valid entity upon entering the

I'm just not a fan of mixing different data type ID (int) and UUID just for the sake of having an "ID". It complicates things too much. 

> 7. The fully formed valid entity is returned to the application later where it can be serialized and presented to the requester.

I would create use case specific data objects instead.

Reply:
> I also don't want to manage different API versions (for different structures) within entities. Entities are just not the right "tool" for this task (IMHO) and looks like a poor design.

API versioning is a pretty widely accepted concept at this point. 

> I would validate the raw data (array) first and then create the DTO from it. The reason is, that the incoming data are casted into specific data types. If you validate casted data, your validation does't give you correct results.

I take back what I said about validating against a DTO, one should validate against the incoming raw data, I agree. I personally like to use JSON Schema for this. 

> DTOs are commonly known as objects passing through the architectural boundaries to transfer data (and not entities).

I'm aware. Again, this is a design choice. Clearly you prefer DTOs over fully formed entities, which is where I disagree.

> If you have an entity then you imply that you always have an ID. This makes no sense for me in every case. A DTO is for transfering data inside and outside the domain (not an entity).

This is what UUIDs are for. Then you don't have to rely on your persistence layer to manager identity. 

> I'm just not a fan of mixing different data type ID (int) and UUID just for the sake of having an "ID". It complicates things too much.

And you don't believe that having varying identity properties in a response object when creating a resource complicates things too much? 

We have different design philosophies and that's fine. It's going to be tough for me to approve of this tutorial since it's going to be used as a reference point for years to come with design patterns that I disagree with. I'm not sure what to do at this point.

Reply:
Okay, I understand that the hexagonal architecture isn't necessarily yours. It is important to me that you should also stand behind it due to later questions. For this reason I am compelled to close this PR. Thank you very much for your valuable feedback.

Reply:
@odan please don't take this the wrong way. I appreciate all the efforts you've been putting recently. I think that I probably will need to make the tutorial in line with the current Slim-Skeleton repo so it's easier to follow along.